### PR TITLE
Add tests for mixed precision training and autocast

### PR DESCRIPTION
**Pull request by assassin1991**
<details>
  <summary><b>Add tests for mixed precision training and autocast</b></summary>

- [ ] Was this discussed/agreed via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?
Adds tests for mixed precision training and autocast

Fixes # (issue).

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in GitHub issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/test_pytorch/test_precision.py [hunk 1]
```diff
@@ -55,16 +90,159 @@ def setup(self, trainer: Trainer, pl_module: LightningModule, stage: str) -> Non
...
 def test_unsupported_precision_plugin():
+    """Tests unsupported HPUPrecisionPlugin init."""
     with pytest.raises(ValueError, match=r"accelerator='hpu', precision='mixed'\)` is not supported."):
         HPUPrecisionPlugin(precision="mixed")
+
+
+class TestMixedPrecisionHPU:
+    """Test mixed precision training methods for HPU."""
+
+    class BaseBM(BoringModel):
```
[the remark this answers is not in the document]
Comment (assassin1991) [reply tests/test_pytorch/test_precision.py]: moved models and removed test class altogether

File: CHANGELOG.md [hunk 2]
```diff
@@ -39,6 +39,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added Pytorch HPU profiler support ([#15](https://github.com/Lightning-AI/lightning-Habana/pull/15))
 - Added basic HPU infra support for deep speed ([#21](https://github.com/Lightning-AI/lightning-Habana/pull/21))
 - Added Pytorch HPU datamodule support ([#16](https://github.com/Lightning-AI/lightning-Habana/pull/16))
+- Added tests for mixed precision training ([#36](https://github.com/Lightning-AI/lightning-Habana/pull/36))
```
Comment (jerome-habana) [inline CHANGELOG.md:42]: this should goto unreleased section

File: docs/source/intermediate.rst [hunk 3]
```diff
@@ -69,6 +73,53 @@ The following is an excerpt from an MNIST example implemented on a single HPU.
...
+
+For more granular control over with mixed precision training, one can use torch.autocast from native PyTorch.
+
+Instances of autocast serve as context managers or decorators that allow regions of your script to run in mixed precision.
+These also allow for fine tuning with `enabled` for enabling and disabling mixed precision training for certain parts of the code.
+
+.. code-block:: python
+
+    import torch
+    import pytorch_lightning as pl
```
Comment (jerome-habana) [inline docs/source/intermediate.rst:87]: use lightning directly as in other documents

File: tests/test_pytorch/test_precision.py [hunk 4]
```diff
@@ -37,12 +49,106 @@ def hmp_params(request):
...
+        self.log("val_loss", loss.get("x").to(torch.bfloat16), prog_bar=True, sync_dist=True)
+        return loss
+
+
+class BaseBMActive(BaseBM):
+    """Model with torch.autocast enabled check."""
+
+    def forward(self, x):
+        """Forward."""
+        assert torch.hpu.is_autocast_hpu_enabled()
```
Comment (jerome-habana) [inline tests/test_pytorch/test_precision.py:107]: can this be moved to plugin ?

File: tests/test_pytorch/test_precision.py [hunk 5]
```diff
@@ -55,16 +161,152 @@ def setup(self, trainer: Trainer, pl_module: LightningModule, stage: str) -> Non
...
+                # Re-entering autocast enabled region
+                assert torch.hpu.is_autocast_hpu_enabled()
+                x = torch.mm(x, identity)
+                assert x.dtype == torch.bfloat16
+            return self.layer(x)
+
+    assert run_training(tmpdir, BMAutocastGranularControl, []) is not None
+
+
+@pytest.mark.xfail(reason="Env needs to be set")
```
Comment (jerome-habana) [inline tests/test_pytorch/test_precision.py:289]: can we add the test in pipeline?

**Review by jerome-habana — APPROVED** [review]

**Pull request merged**